igvc_navigation/src/back_up_recovery/back_up_recovery.cpp

```diff
@@ -0,0 +1,100 @@
+#include "back_up_recovery.h"
+#include <cmath>
+#include <pluginlib/class_list_macros.h>
+#include <geometry_msgs/Twist.h>
+#include <algorithm>
+#include <igvc_utils/NodeUtils.hpp>
+#include <tf2/utils.h>
+
+// register this planner as a RecoveryBehavior plugin
+PLUGINLIB_EXPORT_CLASS(back_up_recovery::BackUpRecovery, nav_core::RecoveryBehavior);
+
+namespace back_up_recovery
+{
+BackUpRecovery::BackUpRecovery() : local_costmap_(nullptr), initialized_(false), world_model_(nullptr)
+{
+}
+
+void BackUpRecovery::initialize(std::string name, tf2_ros::Buffer *, costmap_2d::Costmap2DROS *,
+                                costmap_2d::Costmap2DROS *local_costmap)
+{
+  if (!initialized_)
+  {
+    local_costmap_ = local_costmap;
+
+    // get some parameters from the parameter server
+    ros::NodeHandle private_nh("~");
+
+    private_nh.getParam("/move_base_flex/back_up_recovery/velocity", velocity_);
+    private_nh.getParam("/move_base_flex/back_up_recovery/frequency", frequency_);
+    private_nh.getParam("/move_base_flex/back_up_recovery/threshold_distance",
+                        threshold_distance_);  // distance robot backs up before exiting back up behavior
+    private_nh.getParam("/move_base_flex/back_up_recovery/obstacle_distance",
+                        obstacle_distance_);  // closest the robot will get to an obstacle from behind
+
+    world_model_ = new base_local_planner::CostmapModel(*local_costmap_->getCostmap());
+
+    initialized_ = true;
+  }
+  else
+  {
+    ROS_ERROR("You should not call initialize twice on this object, doing nothing");
+  }
+}
+
+BackUpRecovery::~BackUpRecovery()
+{
+  delete world_model_;
+}
+
+void BackUpRecovery::runBehavior()
+{
+  if (!initialized_)
+  {
+    ROS_ERROR("This object must be initialized before runBehavior is called");
+    return;
+  }
+
+  if (local_costmap_ == nullptr)
+  {
+    ROS_ERROR("The costmap passed to the BackupRecovery object cannot be NULL. Doing nothing.");
+    return;
+  }
+  ROS_WARN("Backup recovery behavior started.");
+
+  ros::Rate r(frequency_);
+  ros::NodeHandle n;
+  ros::Publisher vel_pub = n.advertise<geometry_msgs::Twist>("cmd_vel", 5);
+
+  geometry_msgs::PoseStamped global_pose;
+  local_costmap_->getRobotPose(global_pose);
```

Comment:
How does getRobotPose() work?

Reply:
I don't know dude, it just gets the pose and it works.

Reply:
We tell it which topic to get the robot pose from, ie. `/odometry/filtered`, and then this is just a method that gets the robot pose from the "base" costmap, which already subscribes to this topic.